File: 0_1.baekjoon/brute-force/java/src/BOJ9663.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class BOJ9663 {
    private static int N;
    private static int[] column;
    private static int answer = 0;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(br.readLine());

        column = new int[N];

        System.out.println(solve());
    }

    private static int solve() {

        dfs(0);

        return answer;
    }

    private static void dfs(int depth) {
        if (depth == N) {
            answer++;
            return;
        }

        for (int i=0;i<N;i++) {
            if (check(depth, i)) {
                column[depth] = i;
                dfs(depth + 1);
                column[depth] = 0;
            }
        }
    }

    private static boolean check(int row, int col) {
        for (int i=0;i<row;i++) {
            if (col == column[i])
                return false;

            if ((row - i) == Math.abs(col - column[i])) {
                return false;
            }

        }
        return true;
    }
}

File: 0_1.baekjoon/brute-force/kotlin/src/BOJ9663.kt
import kotlin.math.abs

fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()
    val column = IntArray(n) { 0}

    var answer = 0

    fun check(row: Int, col: Int): Boolean {
        for (i in 0 until row) {
            if (col == column[i]) {
                return false
            }

            if ((row - i) == abs(col - column[i])) {
                return false
            }
        }
        return true
    }

    fun dfs(depth: Int) {
        if (depth == n) {
            answer++
            return
        }

        for (i in 0 until n) {
            if (check(depth, i)) {
                column[depth] = i
                dfs(depth + 1)
                column[depth] = 0
            }
        }
    }
    dfs(0)

    print(answer)

}